exonum/src/runtime/migrations.rs

```diff
@@ -0,0 +1,485 @@
+///
+/// The support of prerelease versions needs to be explicitly enabled by using the
+/// [`with_prereleases`](#method.with_prereleases) constructor. Otherwise, a prerelease mentioned
+/// in the builder stage will lead to a panic, and [`select`](#method.select) will return an error
+/// if a prerelease is specified as a `start_version`.
+#[derive(Debug)]
+pub struct LinearMigrations {
+    min_start_version: Option<Version>,
+    latest_version: Version,
+    scripts: BTreeMap<Version, MigrationScript>,
+    support_prereleases: bool,
+}
+
+impl LinearMigrations {
+    /// Creates a new set of migrations with the specified latest supported version.
+    pub fn new(latest_version: Version) -> Self {
+        assert!(
+            !latest_version.is_prerelease(),
+            "Prerelease versions require using `with_prereleases` constructor"
+        );
+        Self {
+            support_prereleases: false,
+            ..Self::with_prereleases(latest_version)
+        }
+    }
+
+    /// Creates a new set of migrations with the specified latest supported version.
+    ///
+    /// Unlike the `new` constructor, this one indicates that prerelease versions are allowed
+    /// in the list of migrations and as the start version.
+    pub fn with_prereleases(latest_version: Version) -> Self {
+        Self {
+            min_start_version: None,
+            latest_version,
+            scripts: BTreeMap::new(),
+            support_prereleases: true,
+        }
+    }
+
+    fn check_prerelease(&self, version: &Version) {
+        if !self.support_prereleases {
+            assert!(
+                !version.is_prerelease(),
+                "Prerelease versions require using `with_prereleases` constructor"
+            );
+        }
+    }
+
+    /// Signals to return an error if the starting version is less than the specified version.
+    pub fn forget_before(mut self, version: Version) -> Self {
+        self.check_prerelease(&version);
+        self.min_start_version = Some(version);
+        self
+    }
+
+    /// Adds a migration script at the specified version.
+    pub fn migrate<F>(mut self, version: Version, script: F) -> Self
+    where
+        F: FnOnce(&mut MigrationContext) + Send + 'static,
+    {
+        self.check_prerelease(&version);
+        assert!(
+            version <= self.latest_version,
+            "Cannot add a script for a future version {} (the latest version is {})",
+            version,
+            self.latest_version
+        );
+        let script = MigrationScript::new(script, version.clone());
+        self.scripts.insert(version, script);
+        self
+    }
+
+    /// Selects a list of migration scripts based on the provided start version of the artifact.
+    pub fn select(
+        self,
+        start_version: &Version,
+    ) -> Result<Vec<MigrationScript>, DataMigrationError> {
+        if !self.support_prereleases && start_version.is_prerelease() {
```

Comment:
In `check_prerelease` similar check is an assertion, and here we just return an error. I think it'll be better to have a consistent reactions.

Reply:
This is intentional. Using a prerelease version in in `migrate` is the decision made by the service developer, so we want to inform that he is wrong in his ways as quickly and as prominently as possible. (Ideally, during compilation time, but I didn't come up with the way to do that just yet, although it seems possible via with proc macros.)

In contrast, here `start_version` is *not* chosen by the service developer, so panicking would be inappropriate.